components/core/src/clp_s/JsonParser.cpp

```diff
@@ -520,6 +521,317 @@ bool JsonParser::parse() {
+                                      .get_immutable_view<clp::ir::FourByteEncodedTextAst>()
+                                      .decode_and_unparse()
+                                      .value();
+                    encoded_str = clp::ffi::validate_and_escape_utf8_string(decodedValue.c_str())
+                                          .value();
+                }
+                m_current_parsed_message.add_value(node_id, encoded_str);
+            } break;
+            case NodeType::UnstructuredArray: {
+                std::string array_str;
+                if (pair.second.value().is<clp::ir::EightByteEncodedTextAst>()) {
+                    array_str = pair.second.value()
+                                        .get_immutable_view<clp::ir::EightByteEncodedTextAst>()
+                                        .decode_and_unparse()
+                                        .value();
+                } else {
+                    array_str = pair.second.value()
+                                        .get_immutable_view<clp::ir::FourByteEncodedTextAst>()
+                                        .decode_and_unparse()
+                                        .value();
+                }
+                m_current_parsed_message.add_value(node_id, array_str);
+                break;
+            }
+            default:
+                // Don't need to add value for obj or null
+                break;
+        }
+        m_current_schema.insert_ordered(node_id);
+    }
+
+    int32_t current_schema_id = m_archive_writer->add_schema(m_current_schema);
+    m_current_parsed_message.set_id(current_schema_id);
+    m_archive_writer->append_message(current_schema_id, m_current_schema, m_current_parsed_message);
+    return;
+}
+
+bool JsonParser::parse_from_IR() {
+    std::map<std::tuple<int, NodeType>, int> id_conversion_cache;
+    m_archive_writer->add_node(-1, NodeType::Unknown, "root");
+
+    for (auto& file_path : m_file_paths) {
+        std::cout << file_path << std::endl;
+        std::vector<char> ir_buf;
+        char temp_ir_buf[10'000];
+        FileReader infile;
+        infile.open(file_path);
+        if (false == infile.is_open()) {
+            m_archive_writer->close();
+            return false;
+        }
+        int fsize = std::filesystem::file_size(file_path);
+        if (0 == fsize) {
+            m_archive_writer->close();
+            return false;
+        }
+        ZstdDecompressor zd;
+        zd.open(infile, fsize);
+        size_t num_bytes_read = 0;
+        do {
+            num_bytes_read = 0;
+            zd.try_read(temp_ir_buf, 10'000, num_bytes_read);
+            if (num_bytes_read != 0) {
+                ir_buf.insert(ir_buf.end(), temp_ir_buf, temp_ir_buf + num_bytes_read);
+            }
+        } while (num_bytes_read == 10'000);
+        zd.close();
+        infile.close();
+
+        BufferReader reader{size_checked_pointer_cast<char>(ir_buf.data()), ir_buf.size()};
```

Comment:
We don't need to read the entire file into a memory buffer for file ingestion and decompress. Instead, we can open a `clp::FileReader` inside a `clp::streaming_compression::zstd::Decompressor` to automate disk IO and zstd decompression. The deserializer takes a `ReaderInterface`, so an instance of any derived class can be a valid input.